{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Report if residues interacting with a specific chain have equivalent residues in an hhsuite-generated alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook is meant to be used when you are studing a complex that contains at least two or more proteins because it is meant to generate a report on the residues of a protein chain of interest that interacts with another in a structure and tell you what parts of the chain of interest have equivalents in a remote homolog. This notebook illustrates bringing together the structural data from a released structure with a remote homology alignment by tools from HH-Suite3. The idea is you may have a protein where the structure has been solved and if that chain has been used to identify remote homologs, you may be curious about whether residues involved in interacting with other chains in the complex are found in the identified homolog and what are the identities.\n",
    "\n",
    "This effort uses information from the PDBsum report for the corresponding chains in a specified structure. See [pdbsum-binder: demonstrating analysis of PDBsum-related data via active Jupyter sessions provided via MyBinder.org](https://github.com/fomightez/pdbsum-binder) and [my collection of utility scripts for working with data from PDBsum](https://github.com/fomightez/structurework/tree/master/pdbsum-utilities) for more resources abut PDBsum and using it in conjuction with Jupyter/Python.\n",
    "\n",
    "The portion dealing with getting the protein-protein interaction data, and some of the general apprach to this example applying to one protein protein interaction as the initial example, is based on my notebook [Using PDBsum data to highlight changes in protein-protein interactions.ipynb](https://nbviewer.jupyter.org/github/fomightez/pdbsum-binder/blob/main/notebooks/Using%20PDBsum%20data%20to%20highlight%20changes%20in%20protein-protein%20interactions.ipynb) that is available and ready to be run in active Binder sessions launched from [here](https://github.com/fomightez/pdbsum-binder).  The idea is to expand on that for when you want to look at multiple structure examples for the involved chain and that follow-up will be based on [Using snakemake to highlight changes in multiple protein-protein interactions via PDBsum data.ipynb](https://nbviewer.jupyter.org/github/fomightez/pdbsum-binder/blob/main/notebooks/Using%20snakemake%20to%20highlight%20changes%20in%20multiple%20protein-protein%20interactions%20via%20PDBsum%20data.ipynb) available as an active running file in Binder sessions launched from [the same place](https://github.com/fomightez/pdbsum-binder).\n",
    "\n",
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get necessary scripts\n",
    "\n",
    "Get the `hhsuite3_results_to_df.py` script that will be used to mine the results file(s) generated by HH-suite3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "file_needed = \"hhsuite3_results_to_df.py\"\n",
    "if not os.path.isfile(file_needed):\n",
    "    !curl -OL https://raw.githubusercontent.com/fomightez/sequencework/master/hhsuite3-utilities/hhsuite3_results_to_df.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the `pdbsum_prot_interactions_list_to_df.py` script that will be used to collect interaction data for a protein pair from [PDBsum](http://www.ebi.ac.uk/thornton-srv/databases/cgi-bin/pdbsum/GetPage.pl?pdbcode=index.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "file_needed = \"pdbsum_prot_interactions_list_to_df.py\"\n",
    "if not os.path.isfile(file_needed):\n",
    "    !curl -OL https://raw.githubusercontent.com/fomightez/structurework/master/pdbsum-utilities/pdbsum_prot_interactions_list_to_df.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the `make_report_on_equivalents_of_interacting_residues.py` script that will be used to make the report on the equivalent residues in the structure that interact with other chains and the aligned sequence of the idenitified homolog."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "file_needed = \"make_report_on_equivalents_of_interacting_residues.py\"\n",
    "if not os.path.isfile(file_needed):\n",
    "    !curl -OL https://raw.githubusercontent.com/fomightez/sequencework/master/hhsuite3-utilities/make_report_on_equivalents_of_interacting_residues.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get example input data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get an example file  HH-suite3-generated results.  \n",
    "\n",
    "(You can use your own `.hhr` file if you have one and the chain used in the query has a corresponding structure; however, you'll need to change the entries in this demo notebook to correspond.)\n",
    "\n",
    "For ease, I'll just some of the results files that the biopython package uses to test the parsing code is working well. They are publically available [here](https://github.com/biopython/biopython/tree/master/Tests/HHsuite); the `README` file [there](https://github.com/biopython/biopython/tree/master/Tests/HHsuite) summarizes the content of those files.\n",
    "\n",
    "If you prefer to use your own result files, upload your file to this session and change the appropriate file names in later steps.  \n",
    "(Note for a couple of the `.hhr` files, I noticed they had `Done!` close to the last line of the file; however, I hadn't seen that in the `.hhr` files from HH-suite3 that I developed with and it caused errors for my `hhsuite3_results_to_df.py`. To make them more consistent, I'm removing that, assuming it is from format version of HH-suite since biopython which is a source of these test files mentions that version in the current documentation.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "file_needed = \"results_S288C_NOP1.hhr\"\n",
    "if not os.path.isfile(file_needed):\n",
    "    !curl -OL https://gist.githubusercontent.com/fomightez/cbdca72f5990146170f6d15789234dfb/raw/92f6ad30056b4c930d480b6da69265b0101e0cc6/results_S288C_NOP1.hhr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll see how to specify use of this later. You'd obviously want to replace use of this with your own HH-suite3-generated results to adapt this notebook to your own interests."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Specifying particulars for the main analysis\n",
    "\n",
    "### Getting the interactions file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The portion dealing with getting the protein-protein interaction data, and some of the general apprach to this example applying to one protein protein interaction as the initial example, is based on my notebook [Using PDBsum data to highlight changes in protein-protein interactions.ipynb](https://nbviewer.jupyter.org/github/fomightez/pdbsum-binder/blob/main/notebooks/Using%20PDBsum%20data%20to%20highlight%20changes%20in%20protein-protein%20interactions.ipynb) that is available and ready to be run in active Binder sessions launched from [here](https://github.com/fomightez/pdbsum-binder). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's define the structure to use. (We'll use [6zdt](https://www.rcsb.org/structure/6ZDT) for the demo.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "structure = \"6zdt\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's define the chains to use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "structure_chain1 = \"A\"\n",
    "structure_chain2 = \"B\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Chain `A` in this structure is baker's yeast Nop1p, a.k.a. fibrillarin in other eukaryotic organisms.\n",
    "Importantly, **chain1 here will be the one we need to correspond to the chain represented as the query chain the HH-suite-generated data.**  \n",
    "If chain `B` in the structure had been the one used in the query and the other chain we wanted to examine it's interactions with was `X`, we would have made the code like this:\n",
    "\n",
    "```python\n",
    "structure_chain1 = \"B\"\n",
    "structure_chain2 = \"X\"\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, let's define what we want the data files saved as:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "structure_chains_data_name = \"datai_6zdt_A_B.txt\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Code based on first notebook in the pdbsum-binder launches to get the interaction data files for each protein pair for each of the two structures:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 16699    0 16673  100    26  15241     23  0:00:01  0:00:01 --:--:-- 15278\n"
     ]
    }
   ],
   "source": [
    "def get_protein_inter_data_files(pdb_code,chain1,chain2,output_file_name):\n",
    "    '''\n",
    "    Takes a PDB entry accession identifier alphanumeic (PDB code), a chain \n",
    "    identifier for chain 1 and chain identifier for chain 2, along with a\n",
    "    name to give the file produced when the data is retrieved and saved.\n",
    "\n",
    "    The proteins have to interact in the structure for meaningful data to be returned.\n",
    "    '''\n",
    "    source_url = \"http://www.ebi.ac.uk/thornton-srv/databases/cgi-bin/pdbsum/GetIface.pl\"\n",
    "    !curl -L -o {output_file_name} --data \"pdb={pdb_code.lower()}&chain1={chain1}&chain2={chain2}\" {source_url}\n",
    "# Get data file for the structure\n",
    "get_protein_inter_data_files(structure,structure_chain1,structure_chain2,structure_chains_data_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can list the files present to verify the data file was obtained."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " datai_6zdt_A_B.txt\n",
      " Encephalitozoon_cuniculi_ecuniii_l_gca_001078035.ECIIIL.pep.all.fa.gz\n",
      " hhs3_results_pickled_df.pkl\n",
      " hhsuite3_results_to_df.py\n",
      " make_report_on_equivalents_of_interacting_residues.py\n",
      " pdbsum_prot_interactions_list_to_df.py\n",
      " prot_int_pickled_df.pkl\n",
      " \u001b[0m\u001b[01;34m__pycache__\u001b[0m/\n",
      "'Report if residues interacting with a specific chain have equivalent residues in an hhsuite-generated alignment.ipynb'\n",
      " results_S288C_NOP1.hhr\n",
      "\u001b[01;32m'Using Python to examine HH-suite3 results.ipynb'\u001b[0m*\n",
      "'Using Python to examine HH-suite results - example making multi-fasta sequence format from query and hit sequences.ipynb'\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`datai_6zdt_A_B.txt` will show in the list if all is okay.\n",
    "\n",
    "Next we want to read in the data to make a useful dataframe from it using the main function of the script  `pdbsum_prot_interactions_list_to_df.py`that we retrieved earlier. (See 'Making a Pandas dataframe from the interactions file directly in Jupyter' in `Working with PDBsum in Jupyter Basics.ipynb` that is available and ready to be run in active Binder sessions launched from [here](https://github.com/fomightez/pdbsum-binder) to follow what is going on in this step. That notebook is available [here](https://nbviewer.jupyter.org/github/fomightez/pdbsum-binder/blob/main/notebooks/Working%20with%20PDBsum%20in%20Jupyter%20Basics.ipynb#Making-a-Pandas-dataframe-from-the-interactions-file-directly-in-Jupyter) in static form from [the pdbsum-binder repo](https://github.com/fomightez/pdbsum-binder).)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Provided interactions data read and converted to a dataframe...\n",
      "\n",
      "A dataframe of the data has been saved as a file\n",
      "in a manner where other Python programs can access it (pickled form).\n",
      "RESULTING DATAFRAME is stored as ==> 'prot_int_pickled_df.pkl'\n",
      "\n",
      "Returning a dataframe with the information as well."
     ]
    }
   ],
   "source": [
    "from pdbsum_prot_interactions_list_to_df import pdbsum_prot_interactions_list_to_df\n",
    "i_df = pdbsum_prot_interactions_list_to_df(\"datai_6zdt_A_B.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can look at the initial lines of data in the dataframe to make sure that worked by running the next cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Atom1 no.</th>\n",
       "      <th>Atom1 name</th>\n",
       "      <th>Atom1 Res name</th>\n",
       "      <th>Atom1 Res no.</th>\n",
       "      <th>Atom1 Chain</th>\n",
       "      <th>Atom2 no.</th>\n",
       "      <th>Atom2 name</th>\n",
       "      <th>Atom2 Res name</th>\n",
       "      <th>Atom2 Res no.</th>\n",
       "      <th>Atom2 Chain</th>\n",
       "      <th>Distance</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1219</td>\n",
       "      <td>NZ</td>\n",
       "      <td>LYS</td>\n",
       "      <td>169</td>\n",
       "      <td>A</td>\n",
       "      <td>6012</td>\n",
       "      <td>OE1</td>\n",
       "      <td>GLU</td>\n",
       "      <td>148</td>\n",
       "      <td>B</td>\n",
       "      <td>2.97</td>\n",
       "      <td>Hydrogen bonds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1601</td>\n",
       "      <td>OH</td>\n",
       "      <td>TYR</td>\n",
       "      <td>195</td>\n",
       "      <td>A</td>\n",
       "      <td>6061</td>\n",
       "      <td>OE1</td>\n",
       "      <td>GLN</td>\n",
       "      <td>151</td>\n",
       "      <td>B</td>\n",
       "      <td>3.08</td>\n",
       "      <td>Hydrogen bonds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1760</td>\n",
       "      <td>NE</td>\n",
       "      <td>ARG</td>\n",
       "      <td>205</td>\n",
       "      <td>A</td>\n",
       "      <td>6266</td>\n",
       "      <td>O</td>\n",
       "      <td>VAL</td>\n",
       "      <td>163</td>\n",
       "      <td>B</td>\n",
       "      <td>3.31</td>\n",
       "      <td>Hydrogen bonds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1763</td>\n",
       "      <td>NH2</td>\n",
       "      <td>ARG</td>\n",
       "      <td>205</td>\n",
       "      <td>A</td>\n",
       "      <td>6266</td>\n",
       "      <td>O</td>\n",
       "      <td>VAL</td>\n",
       "      <td>163</td>\n",
       "      <td>B</td>\n",
       "      <td>3.05</td>\n",
       "      <td>Hydrogen bonds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2038</td>\n",
       "      <td>O</td>\n",
       "      <td>PRO</td>\n",
       "      <td>219</td>\n",
       "      <td>A</td>\n",
       "      <td>6176</td>\n",
       "      <td>OG</td>\n",
       "      <td>SER</td>\n",
       "      <td>159</td>\n",
       "      <td>B</td>\n",
       "      <td>2.71</td>\n",
       "      <td>Hydrogen bonds</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Atom1 no. Atom1 name Atom1 Res name  Atom1 Res no. Atom1 Chain  Atom2 no.  \\\n",
       "0       1219         NZ            LYS            169           A       6012   \n",
       "1       1601         OH            TYR            195           A       6061   \n",
       "2       1760         NE            ARG            205           A       6266   \n",
       "3       1763        NH2            ARG            205           A       6266   \n",
       "4       2038          O            PRO            219           A       6176   \n",
       "\n",
       "  Atom2 name Atom2 Res name  Atom2 Res no. Atom2 Chain  Distance  \\\n",
       "0        OE1            GLU            148           B      2.97   \n",
       "1        OE1            GLN            151           B      3.08   \n",
       "2          O            VAL            163           B      3.31   \n",
       "3          O            VAL            163           B      3.05   \n",
       "4         OG            SER            159           B      2.71   \n",
       "\n",
       "             type  \n",
       "0  Hydrogen bonds  \n",
       "1  Hydrogen bonds  \n",
       "2  Hydrogen bonds  \n",
       "3  Hydrogen bonds  \n",
       "4  Hydrogen bonds  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting the alignment of the yeast chain with the ortholog chain\n",
    "\n",
    "This will be where we get the alignment of the chain represented in the structure with the ortholog in the other organism by extracting the alignment data from the HH-suite3-generated data.\n",
    "\n",
    "We retrieved the results file `results_S288C_NOP1.hhr` earlier. And so we can easily make a dataframe of that using the main function of the script `hhsuite3_results_to_df` we retrieved earlier.  (See the bottom of the notebook `Using Python to examine HH-suite3 results.ipynb` that should be available in this session [or repo] if you need to better understand what is going on in this step.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Provided results read and converted to a dataframe...\n",
      "\n",
      "A dataframe of the data has been saved as a file\n",
      "in a manner where other Python programs can access it (pickled form).\n",
      "RESULTING DATAFRAME is stored as ==> 'hhs3_results_pickled_df.pkl'\n",
      "\n",
      "Returning a dataframe with the information as well."
     ]
    }
   ],
   "source": [
    "from hhsuite3_results_to_df import hhsuite3_results_to_df\n",
    "hh_df = hhsuite3_results_to_df(\"results_S288C_NOP1.hhr\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can look at the initial lines of data in the dataframe to make sure that worked by running the next cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hit_num</th>\n",
       "      <th>qid</th>\n",
       "      <th>qtitle</th>\n",
       "      <th>query_length</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>hid</th>\n",
       "      <th>htitle</th>\n",
       "      <th>hit_length</th>\n",
       "      <th>hstart</th>\n",
       "      <th>...</th>\n",
       "      <th>Similarity</th>\n",
       "      <th>Sum_probs</th>\n",
       "      <th>Template_Neff</th>\n",
       "      <th>size_diff</th>\n",
       "      <th>qseq</th>\n",
       "      <th>hconsensus</th>\n",
       "      <th>qconsensus</th>\n",
       "      <th>hseq</th>\n",
       "      <th>aln_confidence</th>\n",
       "      <th>pw_conservation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>NOP1</td>\n",
       "      <td>NOP1 YDL014W</td>\n",
       "      <td>327</td>\n",
       "      <td>80</td>\n",
       "      <td>326</td>\n",
       "      <td>KMV65234</td>\n",
       "      <td>KMV65234 pep supercontig:ECIIIL:CH10_ECIII-L:9...</td>\n",
       "      <td>291</td>\n",
       "      <td>41</td>\n",
       "      <td>...</td>\n",
       "      <td>1.074</td>\n",
       "      <td>204.2</td>\n",
       "      <td>8.7</td>\n",
       "      <td>36</td>\n",
       "      <td>ARGGAKVVIEPH-RHAGVYIARGKEDLLVTKNMAPGESVYGEKRIS...</td>\n",
       "      <td>~~~~~~~~~~~~~~~~~~~~~~~~~~~~~t~~~~~g~~vyge~~~~...</td>\n",
       "      <td>~~~~~~~~~~~~-~~~~~~~~~~~~~~~~t~~~~~g~~vyge~~~~...</td>\n",
       "      <td>AGLDRKVLVEPHPRFPGVYISRGKEDLLLTRNLVPGVSVYGEKRVA...</td>\n",
       "      <td>4567899999999998888899999999999999999988753356...</td>\n",
       "      <td>.++++++.|+||++.+.+.++.+|||+|||+|.++..++...++.....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>NOP1</td>\n",
       "      <td>NOP1 YDL014W</td>\n",
       "      <td>327</td>\n",
       "      <td>165</td>\n",
       "      <td>272</td>\n",
       "      <td>KMV66152</td>\n",
       "      <td>KMV66152 pep supercontig:ECIIIL:CH05_ECIII-L:1...</td>\n",
       "      <td>247</td>\n",
       "      <td>46</td>\n",
       "      <td>...</td>\n",
       "      <td>0.085</td>\n",
       "      <td>69.5</td>\n",
       "      <td>9.1</td>\n",
       "      <td>80</td>\n",
       "      <td>APGKKVLYLGAASGTSVSHVSDVVGPEGVVYAVEFSHRPGRELISM...</td>\n",
       "      <td>~~~~~vLDiGcG~G~~~~~l~~~---g~~v~gvD~s~~~i~~a~~~...</td>\n",
       "      <td>~~~~~VLDiG~G~G~~~~~la~~~~~~~~v~gvD~s~~~i~~a~~~...</td>\n",
       "      <td>KDGGLVLDVGCGSGLSGSVLSES---GYPWIGVDISMEMLKLGMER...</td>\n",
       "      <td>45678999999999998888765432</td>\n",
       "      <td>.++.+|||+|||+|.++..++...+.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>NOP1</td>\n",
       "      <td>NOP1 YDL014W</td>\n",
       "      <td>327</td>\n",
       "      <td>165</td>\n",
       "      <td>272</td>\n",
       "      <td>KMV65418</td>\n",
       "      <td>KMV65418 pep supercontig:ECIIIL:CH09d_ECIII-L:...</td>\n",
       "      <td>210</td>\n",
       "      <td>43</td>\n",
       "      <td>...</td>\n",
       "      <td>0.078</td>\n",
       "      <td>63.8</td>\n",
       "      <td>12.4</td>\n",
       "      <td>117</td>\n",
       "      <td>APGKKVLYLGAASGTSVSHVSDVVGPEGVVYAVEFSHRPGRELISM...</td>\n",
       "      <td>~~~~~vldiG~G~G~~~~--------~~~~~g~d~~~~~~~~~~~~...</td>\n",
       "      <td>~~~~~VLDiG~G~G~~~~~la~~~~~~~~v~gvD~s~~~i~~a~~~...</td>\n",
       "      <td>TRESIVLDAGCGNGRSFL--------VPCMVGMDYCLGLLNDARAA...</td>\n",
       "      <td>4567999999999964335432</td>\n",
       "      <td>.++.+|||+|||+|.....+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>NOP1</td>\n",
       "      <td>NOP1 YDL014W</td>\n",
       "      <td>327</td>\n",
       "      <td>166</td>\n",
       "      <td>300</td>\n",
       "      <td>KMV65957</td>\n",
       "      <td>KMV65957 pep supercontig:ECIIIL:CH06_ECIII-L:7...</td>\n",
       "      <td>164</td>\n",
       "      <td>23</td>\n",
       "      <td>...</td>\n",
       "      <td>0.050</td>\n",
       "      <td>76.0</td>\n",
       "      <td>10.4</td>\n",
       "      <td>163</td>\n",
       "      <td>PGKKVLYLGAASGTSVSHVSDVVGPEGVVYAVEFSHRPGRELISMA...</td>\n",
       "      <td>~~~~vld~gcG~G~~~~~l~~~----~~v~~~D~~~~~~~~~~~--...</td>\n",
       "      <td>~~~~VLDiG~G~G~~~~~la~~~~~~~~v~gvD~s~~~i~~a~~~~...</td>\n",
       "      <td>EMKIVLDLGTSTGVITEQLRKR----NTVVSTDLNIRALESHRG--...</td>\n",
       "      <td>4678999999999998877655531</td>\n",
       "      <td>++.+|||+|||+|.+...+...+..</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>NOP1</td>\n",
       "      <td>NOP1 YDL014W</td>\n",
       "      <td>327</td>\n",
       "      <td>163</td>\n",
       "      <td>273</td>\n",
       "      <td>KMV65189</td>\n",
       "      <td>KMV65189 pep supercontig:ECIIIL:CH10_ECIII-L:2...</td>\n",
       "      <td>283</td>\n",
       "      <td>46</td>\n",
       "      <td>...</td>\n",
       "      <td>0.177</td>\n",
       "      <td>71.5</td>\n",
       "      <td>11.4</td>\n",
       "      <td>44</td>\n",
       "      <td>FIAPGKKVLYLGAASGTSVSHVSDVVGPEGVVYAVEFSHRPGRELI...</td>\n",
       "      <td>~~~~~~~vldiG~G~G~~~~~~~~~~--~~~v~~~d~~~~~~~~~~...</td>\n",
       "      <td>~~~~~~~VLDiG~G~G~~~~~la~~~~~~~~v~gvD~s~~~i~~a~...</td>\n",
       "      <td>YTKRGDSVLDLGCGKGGDLLKYERAG--IGEYYGVDIAEVSINDAR...</td>\n",
       "      <td>3456789999999999988777655198876543</td>\n",
       "      <td>...++.+|||+|||.|.++..+....|++...+.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   hit_num   qid        qtitle  query_length  qstart  qend       hid  \\\n",
       "0        1  NOP1  NOP1 YDL014W           327      80   326  KMV65234   \n",
       "1        2  NOP1  NOP1 YDL014W           327     165   272  KMV66152   \n",
       "2        3  NOP1  NOP1 YDL014W           327     165   272  KMV65418   \n",
       "3        4  NOP1  NOP1 YDL014W           327     166   300  KMV65957   \n",
       "4        5  NOP1  NOP1 YDL014W           327     163   273  KMV65189   \n",
       "\n",
       "                                              htitle  hit_length  hstart  ...  \\\n",
       "0  KMV65234 pep supercontig:ECIIIL:CH10_ECIII-L:9...         291      41  ...   \n",
       "1  KMV66152 pep supercontig:ECIIIL:CH05_ECIII-L:1...         247      46  ...   \n",
       "2  KMV65418 pep supercontig:ECIIIL:CH09d_ECIII-L:...         210      43  ...   \n",
       "3  KMV65957 pep supercontig:ECIIIL:CH06_ECIII-L:7...         164      23  ...   \n",
       "4  KMV65189 pep supercontig:ECIIIL:CH10_ECIII-L:2...         283      46  ...   \n",
       "\n",
       "   Similarity  Sum_probs  Template_Neff  size_diff  \\\n",
       "0       1.074      204.2            8.7         36   \n",
       "1       0.085       69.5            9.1         80   \n",
       "2       0.078       63.8           12.4        117   \n",
       "3       0.050       76.0           10.4        163   \n",
       "4       0.177       71.5           11.4         44   \n",
       "\n",
       "                                                qseq  \\\n",
       "0  ARGGAKVVIEPH-RHAGVYIARGKEDLLVTKNMAPGESVYGEKRIS...   \n",
       "1  APGKKVLYLGAASGTSVSHVSDVVGPEGVVYAVEFSHRPGRELISM...   \n",
       "2  APGKKVLYLGAASGTSVSHVSDVVGPEGVVYAVEFSHRPGRELISM...   \n",
       "3  PGKKVLYLGAASGTSVSHVSDVVGPEGVVYAVEFSHRPGRELISMA...   \n",
       "4  FIAPGKKVLYLGAASGTSVSHVSDVVGPEGVVYAVEFSHRPGRELI...   \n",
       "\n",
       "                                          hconsensus  \\\n",
       "0  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~t~~~~~g~~vyge~~~~...   \n",
       "1  ~~~~~vLDiGcG~G~~~~~l~~~---g~~v~gvD~s~~~i~~a~~~...   \n",
       "2  ~~~~~vldiG~G~G~~~~--------~~~~~g~d~~~~~~~~~~~~...   \n",
       "3  ~~~~vld~gcG~G~~~~~l~~~----~~v~~~D~~~~~~~~~~~--...   \n",
       "4  ~~~~~~~vldiG~G~G~~~~~~~~~~--~~~v~~~d~~~~~~~~~~...   \n",
       "\n",
       "                                          qconsensus  \\\n",
       "0  ~~~~~~~~~~~~-~~~~~~~~~~~~~~~~t~~~~~g~~vyge~~~~...   \n",
       "1  ~~~~~VLDiG~G~G~~~~~la~~~~~~~~v~gvD~s~~~i~~a~~~...   \n",
       "2  ~~~~~VLDiG~G~G~~~~~la~~~~~~~~v~gvD~s~~~i~~a~~~...   \n",
       "3  ~~~~VLDiG~G~G~~~~~la~~~~~~~~v~gvD~s~~~i~~a~~~~...   \n",
       "4  ~~~~~~~VLDiG~G~G~~~~~la~~~~~~~~v~gvD~s~~~i~~a~...   \n",
       "\n",
       "                                                hseq  \\\n",
       "0  AGLDRKVLVEPHPRFPGVYISRGKEDLLLTRNLVPGVSVYGEKRVA...   \n",
       "1  KDGGLVLDVGCGSGLSGSVLSES---GYPWIGVDISMEMLKLGMER...   \n",
       "2  TRESIVLDAGCGNGRSFL--------VPCMVGMDYCLGLLNDARAA...   \n",
       "3  EMKIVLDLGTSTGVITEQLRKR----NTVVSTDLNIRALESHRG--...   \n",
       "4  YTKRGDSVLDLGCGKGGDLLKYERAG--IGEYYGVDIAEVSINDAR...   \n",
       "\n",
       "                                      aln_confidence  \\\n",
       "0  4567899999999998888899999999999999999988753356...   \n",
       "1                         45678999999999998888765432   \n",
       "2                             4567999999999964335432   \n",
       "3                          4678999999999998877655531   \n",
       "4                 3456789999999999988777655198876543   \n",
       "\n",
       "                                     pw_conservation  \n",
       "0  .++++++.|+||++.+.+.++.+|||+|||+|.++..++...++.....  \n",
       "1                         .++.+|||+|||+|.++..++...+.  \n",
       "2                             .++.+|||+|||+|.....+..  \n",
       "3                          ++.+|||+|||+|.+...+...+..  \n",
       "4                 ...++.+|||+|||.|.++..+....|++...+.  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hh_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The top hit with the score value much larger than the others contains the alignment in which we are interested. And so we can assign the aligned sequences from there as the ones we'll use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "aligned_query_seq = hh_df['qseq'][0]\n",
    "aligned_hit_seq = hh_df['hseq'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Because the aligned sequences may not start at position #1, that needs to be accounted for to get be able to use the numbers from the structure relative the query sequence. For the aligned hit, the position possibly not being the first residue needs to be accounted for to get accurate numbering for the positions of the equivalents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "start_pos_query_seq = hh_df['qstart'][0]\n",
    "start_pos_hit_seq = hh_df['hstart'][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To verify things look good, we can see some of the initial alignment by running the next cell to print a slice of each and say what position it starts at:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ARGGAKVVIEPH-RHAGVYIARGKEDLLVTKNMAPGESVYGEKRISVEEPSKEDGVPPTK\n",
      "AGLDRKVLVEPHPRFPGVYISRGKEDLLLTRNLVPGVSVYGEKRVAVDLEG-------MK\n",
      "The top sequence starts at position 80.\n",
      "The bottom sequence starts at position 41.\n"
     ]
    }
   ],
   "source": [
    "limit = 60\n",
    "print(aligned_query_seq[:limit])\n",
    "print(aligned_hit_seq[:limit])\n",
    "print(f\"The top sequence starts at position {start_pos_query_seq}.\")\n",
    "print(f\"The bottom sequence starts at position {start_pos_hit_seq}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note the aligned sequences and information on start position, and so things look good to actually do the report generation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setting report options\n",
    "\n",
    "There are some **OPTIONAL** settings for the reports. You can choose not to define these and the default settings within the script will be used. For example, if you delete the next cell, the script will still run and represenative context details for the residue will be shown for residues with equivalents and not displayed for those without equivalents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "details_for_those_with_equivalents = True #default in script is that details of context will be shown; can override here\n",
    "details_for_those_without_equivalents = False #default in script is that details of context will NOT be shown; can override here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main analysis step and report generating\n",
    "\n",
    "Now because above we already defined the variables and dataframe containing the data we want to analyze, the script the next cell runs knows to use all those elements to see if the residues involved in the interaction of chain#1 with chain#2 in the specified structure have equivalents in the ortholog protein from the other organism that isn't baker's yeast."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsing data from PDBsum to identify chain A residues interacting with chain B residues ...\n",
      "\n",
      "Identify any residues that are not in the span that is aligned ...\n",
      "\n",
      "Checking for equivalents of the interacting residues in \n",
      "in the other aligned sequence ...\n",
      "\n",
      "Determination of EQUIVALENTS Completed.\n",
      "\n",
      "*************************************RESULTS************************************\n",
      "The following percent of chain A interacting with chain B residues\n",
      "have equivalents in the other aligned sequence:\n",
      "100.0%\n",
      "--------------------------------------------------------------------------------\n",
      "Percent breakdown of the conservation categories for the interacting\n",
      "residues of chain A that have equivalents in\n",
      "the other aligned sequence:\n",
      "identical: 53.1%\n",
      "strongly conserved: 25.0%\n",
      "weakly conserved: 6.2%\n",
      "not conserved: 15.6%\n",
      "--------------------------------------------------------------------------------\n",
      "The specific residue numbers of the conservation categories for the interacting\n",
      "residues of chain A that have equivalents:\n",
      "identical: 169, 193, 195, 205, 208, 215, 217, 219, 220, 221, 222, 231, 232, 233, 260, 261, 323\n",
      "strongly conserved: 218, 229, 234, 236, 257, 258, 263, 320\n",
      "weakly conserved: 201, 228\n",
      "not conserved: 212, 235, 259, 297, 322\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "DETAILS for chain A residues interacting with chain B residues\n",
      "that have equivalents in the other aligned sequence:\n",
      "\n",
      "residue #169 of chain A:\n",
      "       |\n",
      "q IAPGKKVLYLG\n",
      "h MEPGSKVLYLG\n",
      "\n",
      "query aa:K; hit aa: K; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #193 of chain A:\n",
      "       |\n",
      "q VGPEGVVYAVE\n",
      "h VGKDGVVYAVE\n",
      "\n",
      "query aa:V; hit aa: V; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #195 of chain A:\n",
      "       |\n",
      "q PEGVVYAVEFS\n",
      "h KDGVVYAVEFS\n",
      "\n",
      "query aa:Y; hit aa: Y; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #201 of chain A:\n",
      "       |\n",
      "q AVEFSHRPGRE\n",
      "h AVEFSERSGRD\n",
      "\n",
      "query aa:H; hit aa: E; \n",
      "conservation class: weakly conserved\n",
      "\n",
      "\n",
      "residue #205 of chain A:\n",
      "       |\n",
      "q SHRPGRELISM\n",
      "h SERSGRDLINM\n",
      "\n",
      "query aa:R; hit aa: R; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #208 of chain A:\n",
      "       |\n",
      "q PGRELISMAKK\n",
      "h SGRDLINMSMK\n",
      "\n",
      "query aa:I; hit aa: I; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #212 of chain A:\n",
      "       |\n",
      "q LISMAKKRPNI\n",
      "h LINMSMKRPNI\n",
      "\n",
      "query aa:K; hit aa: M; \n",
      "conservation class: not similar\n",
      "\n",
      "\n",
      "residue #215 of chain A:\n",
      "       |\n",
      "q MAKKRPNIIPI\n",
      "h MSMKRPNIVPI\n",
      "\n",
      "query aa:P; hit aa: P; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #217 of chain A:\n",
      "       |\n",
      "q KKRPNIIPIIE\n",
      "h MKRPNIVPIIE\n",
      "\n",
      "query aa:I; hit aa: I; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #218 of chain A:\n",
      "       |\n",
      "q KRPNIIPIIED\n",
      "h KRPNIVPIIED\n",
      "\n",
      "query aa:I; hit aa: V; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #219 of chain A:\n",
      "       |\n",
      "q RPNIIPIIEDA\n",
      "h RPNIVPIIEDA\n",
      "\n",
      "query aa:P; hit aa: P; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #220 of chain A:\n",
      "       |\n",
      "q PNIIPIIEDAR\n",
      "h PNIVPIIEDAR\n",
      "\n",
      "query aa:I; hit aa: I; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #221 of chain A:\n",
      "       |\n",
      "q NIIPIIEDARH\n",
      "h NIVPIIEDARY\n",
      "\n",
      "query aa:I; hit aa: I; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #222 of chain A:\n",
      "       |\n",
      "q IIPIIEDARHP\n",
      "h IVPIIEDARYP\n",
      "\n",
      "query aa:E; hit aa: E; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #228 of chain A:\n",
      "       |\n",
      "q DARHPQKYRML\n",
      "h DARYPSRYRML\n",
      "\n",
      "query aa:Q; hit aa: S; \n",
      "conservation class: weakly conserved\n",
      "\n",
      "\n",
      "residue #229 of chain A:\n",
      "       |\n",
      "q ARHPQKYRMLI\n",
      "h ARYPSRYRMLV\n",
      "\n",
      "query aa:K; hit aa: R; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #231 of chain A:\n",
      "       |\n",
      "q HPQKYRMLIGM\n",
      "h YPSRYRMLVPI\n",
      "\n",
      "query aa:R; hit aa: R; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #232 of chain A:\n",
      "       |\n",
      "q PQKYRMLIGMV\n",
      "h PSRYRMLVPIV\n",
      "\n",
      "query aa:M; hit aa: M; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #233 of chain A:\n",
      "       |\n",
      "q QKYRMLIGMVD\n",
      "h SRYRMLVPIVD\n",
      "\n",
      "query aa:L; hit aa: L; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #234 of chain A:\n",
      "       |\n",
      "q KYRMLIGMVDC\n",
      "h RYRMLVPIVDC\n",
      "\n",
      "query aa:I; hit aa: V; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #235 of chain A:\n",
      "       |\n",
      "q YRMLIGMVDCV\n",
      "h YRMLVPIVDCI\n",
      "\n",
      "query aa:G; hit aa: P; \n",
      "conservation class: not similar\n",
      "\n",
      "\n",
      "residue #236 of chain A:\n",
      "       |\n",
      "q RMLIGMVDCVF\n",
      "h RMLVPIVDCIF\n",
      "\n",
      "query aa:M; hit aa: I; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #257 of chain A:\n",
      "       |\n",
      "q IIALNSHMFLK\n",
      "h IVALNAQYFLK\n",
      "\n",
      "query aa:S; hit aa: A; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #258 of chain A:\n",
      "       |\n",
      "q IALNSHMFLKD\n",
      "h VALNAQYFLKE\n",
      "\n",
      "query aa:H; hit aa: Q; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #259 of chain A:\n",
      "       |\n",
      "q ALNSHMFLKDQ\n",
      "h ALNAQYFLKEG\n",
      "\n",
      "query aa:M; hit aa: Y; \n",
      "conservation class: not similar\n",
      "\n",
      "\n",
      "residue #260 of chain A:\n",
      "       |\n",
      "q LNSHMFLKDQG\n",
      "h LNAQYFLKEGG\n",
      "\n",
      "query aa:F; hit aa: F; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #261 of chain A:\n",
      "       |\n",
      "q NSHMFLKDQGG\n",
      "h NAQYFLKEGGG\n",
      "\n",
      "query aa:L; hit aa: L; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #263 of chain A:\n",
      "       |\n",
      "q HMFLKDQGGVV\n",
      "h QYFLKEGGGVD\n",
      "\n",
      "query aa:D; hit aa: E; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #297 of chain A:\n",
      "       |\n",
      "q KLREERIKPLE\n",
      "h ILRKNSIRPKE\n",
      "\n",
      "query aa:R; hit aa: S; \n",
      "conservation class: not similar\n",
      "\n",
      "\n",
      "residue #320 of chain A:\n",
      "       |\n",
      "q IVVGRYMRSGL\n",
      "h MIIGRFKLSAS\n",
      "\n",
      "query aa:Y; hit aa: F; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #322 of chain A:\n",
      "       |\n",
      "q VGRYMRSGLK\n",
      "h IGRFKLSASE\n",
      "\n",
      "query aa:R; hit aa: L; \n",
      "conservation class: not similar\n",
      "\n",
      "\n",
      "residue #323 of chain A:\n",
      "       |\n",
      "q GRYMRSGLK\n",
      "h GRFKLSASE\n",
      "\n",
      "query aa:S; hit aa: S; \n",
      "conservation class: identical\n",
      "\n",
      "--------------------------------------------------------------------------------"
     ]
    }
   ],
   "source": [
    "%run -i make_report_on_equivalents_of_interacting_residues.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Making the reports more human readable"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note the report that is generated can be made more informative after-the-fact by running the script again and capturing all the output and then using the string `.replace()` method to change the chain designations that come from the PDBsum data to be the protein names and at the same time swap the 'other aligned sequence' for a better phrase to reference the sequence that was aligned to the same chain 1 in the structure. The idea is that the PDBsum data file doesn't have the protein names yet you can easily add them in after to make the report clearer. Plus, you probably have a better way to reference the 'other aligned sequence' than may be found in the HH-suite3-generated alignment. Adding a way to do this when calling the script would add more trouble than it is probably worth since you can perform this trick.    \n",
    "An example is below in the next few cells. When you run the next one, you won't see any output; however, it will be captured."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture out\n",
    "%run -i make_report_on_equivalents_of_interacting_residues.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This next cell will substitute in the protein names and text to replace the 'other aligned seqence', i.e., the 'hit' sequence from the alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsing data from PDBsum to identify fibrillarin (Nop1p) residues interacting with Nop56p residues ...\n",
      "\n",
      "Identify any residues that are not in the span that is aligned ...\n",
      "\n",
      "Checking for equivalents of the interacting residues in \n",
      "in the candidate EC ortholog of fibrillarin ...\n",
      "\n",
      "Determination of EQUIVALENTS Completed.\n",
      "\n",
      "*************************************RESULTS************************************\n",
      "The following percent of fibrillarin (Nop1p) interacting with Nop56p residues\n",
      "have equivalents in the candidate EC ortholog of fibrillarin:\n",
      "100.0%\n",
      "--------------------------------------------------------------------------------\n",
      "Percent breakdown of the conservation categories for the interacting\n",
      "residues of fibrillarin (Nop1p) that have equivalents in\n",
      "the candidate EC ortholog of fibrillarin:\n",
      "identical: 53.1%\n",
      "strongly conserved: 25.0%\n",
      "weakly conserved: 6.2%\n",
      "not conserved: 15.6%\n",
      "--------------------------------------------------------------------------------\n",
      "The specific residue numbers of the conservation categories for the interacting\n",
      "residues of fibrillarin (Nop1p) that have equivalents:\n",
      "identical: 169, 193, 195, 205, 208, 215, 217, 219, 220, 221, 222, 231, 232, 233, 260, 261, 323\n",
      "strongly conserved: 218, 229, 234, 236, 257, 258, 263, 320\n",
      "weakly conserved: 201, 228\n",
      "not conserved: 212, 235, 259, 297, 322\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "DETAILS for fibrillarin (Nop1p) residues interacting with Nop56p residues\n",
      "that have equivalents in the candidate EC ortholog of fibrillarin:\n",
      "\n",
      "residue #169 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q IAPGKKVLYLG\n",
      "h MEPGSKVLYLG\n",
      "\n",
      "query aa:K; hit aa: K; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #193 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q VGPEGVVYAVE\n",
      "h VGKDGVVYAVE\n",
      "\n",
      "query aa:V; hit aa: V; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #195 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q PEGVVYAVEFS\n",
      "h KDGVVYAVEFS\n",
      "\n",
      "query aa:Y; hit aa: Y; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #201 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q AVEFSHRPGRE\n",
      "h AVEFSERSGRD\n",
      "\n",
      "query aa:H; hit aa: E; \n",
      "conservation class: weakly conserved\n",
      "\n",
      "\n",
      "residue #205 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q SHRPGRELISM\n",
      "h SERSGRDLINM\n",
      "\n",
      "query aa:R; hit aa: R; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #208 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q PGRELISMAKK\n",
      "h SGRDLINMSMK\n",
      "\n",
      "query aa:I; hit aa: I; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #212 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q LISMAKKRPNI\n",
      "h LINMSMKRPNI\n",
      "\n",
      "query aa:K; hit aa: M; \n",
      "conservation class: not similar\n",
      "\n",
      "\n",
      "residue #215 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q MAKKRPNIIPI\n",
      "h MSMKRPNIVPI\n",
      "\n",
      "query aa:P; hit aa: P; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #217 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q KKRPNIIPIIE\n",
      "h MKRPNIVPIIE\n",
      "\n",
      "query aa:I; hit aa: I; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #218 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q KRPNIIPIIED\n",
      "h KRPNIVPIIED\n",
      "\n",
      "query aa:I; hit aa: V; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #219 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q RPNIIPIIEDA\n",
      "h RPNIVPIIEDA\n",
      "\n",
      "query aa:P; hit aa: P; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #220 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q PNIIPIIEDAR\n",
      "h PNIVPIIEDAR\n",
      "\n",
      "query aa:I; hit aa: I; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #221 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q NIIPIIEDARH\n",
      "h NIVPIIEDARY\n",
      "\n",
      "query aa:I; hit aa: I; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #222 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q IIPIIEDARHP\n",
      "h IVPIIEDARYP\n",
      "\n",
      "query aa:E; hit aa: E; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #228 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q DARHPQKYRML\n",
      "h DARYPSRYRML\n",
      "\n",
      "query aa:Q; hit aa: S; \n",
      "conservation class: weakly conserved\n",
      "\n",
      "\n",
      "residue #229 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q ARHPQKYRMLI\n",
      "h ARYPSRYRMLV\n",
      "\n",
      "query aa:K; hit aa: R; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #231 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q HPQKYRMLIGM\n",
      "h YPSRYRMLVPI\n",
      "\n",
      "query aa:R; hit aa: R; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #232 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q PQKYRMLIGMV\n",
      "h PSRYRMLVPIV\n",
      "\n",
      "query aa:M; hit aa: M; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #233 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q QKYRMLIGMVD\n",
      "h SRYRMLVPIVD\n",
      "\n",
      "query aa:L; hit aa: L; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #234 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q KYRMLIGMVDC\n",
      "h RYRMLVPIVDC\n",
      "\n",
      "query aa:I; hit aa: V; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #235 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q YRMLIGMVDCV\n",
      "h YRMLVPIVDCI\n",
      "\n",
      "query aa:G; hit aa: P; \n",
      "conservation class: not similar\n",
      "\n",
      "\n",
      "residue #236 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q RMLIGMVDCVF\n",
      "h RMLVPIVDCIF\n",
      "\n",
      "query aa:M; hit aa: I; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #257 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q IIALNSHMFLK\n",
      "h IVALNAQYFLK\n",
      "\n",
      "query aa:S; hit aa: A; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #258 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q IALNSHMFLKD\n",
      "h VALNAQYFLKE\n",
      "\n",
      "query aa:H; hit aa: Q; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #259 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q ALNSHMFLKDQ\n",
      "h ALNAQYFLKEG\n",
      "\n",
      "query aa:M; hit aa: Y; \n",
      "conservation class: not similar\n",
      "\n",
      "\n",
      "residue #260 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q LNSHMFLKDQG\n",
      "h LNAQYFLKEGG\n",
      "\n",
      "query aa:F; hit aa: F; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #261 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q NSHMFLKDQGG\n",
      "h NAQYFLKEGGG\n",
      "\n",
      "query aa:L; hit aa: L; \n",
      "conservation class: identical\n",
      "\n",
      "\n",
      "residue #263 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q HMFLKDQGGVV\n",
      "h QYFLKEGGGVD\n",
      "\n",
      "query aa:D; hit aa: E; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #297 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q KLREERIKPLE\n",
      "h ILRKNSIRPKE\n",
      "\n",
      "query aa:R; hit aa: S; \n",
      "conservation class: not similar\n",
      "\n",
      "\n",
      "residue #320 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q IVVGRYMRSGL\n",
      "h MIIGRFKLSAS\n",
      "\n",
      "query aa:Y; hit aa: F; \n",
      "conservation class: strongly conserved\n",
      "\n",
      "\n",
      "residue #322 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q VGRYMRSGLK\n",
      "h IGRFKLSASE\n",
      "\n",
      "query aa:R; hit aa: L; \n",
      "conservation class: not similar\n",
      "\n",
      "\n",
      "residue #323 of fibrillarin (Nop1p):\n",
      "       |\n",
      "q GRYMRSGLK\n",
      "h GRFKLSASE\n",
      "\n",
      "query aa:S; hit aa: S; \n",
      "conservation class: identical\n",
      "\n",
      "--------------------------------------------------------------------------------"
     ]
    }
   ],
   "source": [
    "sys.stderr.write(out.stderr.replace(\"chain A\",\"fibrillarin (Nop1p)\").replace(\"chain B\",\"Nop56p\").replace(\"other aligned sequence\",\"candidate EC ortholog of fibrillarin\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "You'd substitute the details of your chains of interest and information about the aligned sequence if you were modifying this notebook for your own use and re-running it.\n",
    "\n",
    "----\n",
    "\n",
    "Continue on with the next notebook in the series, [????????](???????). That notebook builds on the ground work here to demonstrate how to use snakemake to scale up this process to produce similar reports on the maintenance or loss of residues equivalent to those involved in interactions with a pair or designated proteins from a structure where the sequence in that structure has been used to identify a remote homolog using HH-sute3 software.\n",
    " \n",
    "----"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
